rustls/src/dns_name.rs

```diff
@@ -0,0 +1,219 @@
+/// DNS name validation according to RFC1035, but with underscores allowed.
+use std::error::Error as StdError;
+use std::fmt;
+
+/// The provided input could not be parsed because
+/// it is not a syntactically-valid DNS Name.
+#[derive(Debug)]
+pub struct InvalidDnsNameError;
+
+impl fmt::Display for InvalidDnsNameError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("invalid dns name")
+    }
+}
+
+impl StdError for InvalidDnsNameError {}
+
+/// A type which encapsulates an owned string that is a syntactically valid DNS name.
+#[derive(Clone, Eq, Hash, PartialEq, Debug)]
+pub struct DnsName(String);
+
+impl<'a> DnsName {
+    /// Produce a borrowed `DnsNameRef` from this owned `DnsName`.
+    pub fn borrow(&'a self) -> DnsNameRef<'a> {
+        DnsNameRef(self.as_ref())
+    }
+
+    /// Validate the given bytes are a DNS name if they are viewed as ASCII.
+    pub fn try_from_ascii(bytes: &[u8]) -> Result<Self, InvalidDnsNameError> {
+        // nb. a sequence of bytes that is accepted by `validate()` is both
+        // valid UTF-8, and valid ASCII.
+        String::from_utf8(bytes.to_vec())
+            .map_err(|_| InvalidDnsNameError)
+            .and_then(Self::try_from)
+    }
+}
+
+impl TryFrom<String> for DnsName {
+    type Error = InvalidDnsNameError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        validate(value.as_bytes())?;
+        Ok(Self(value))
+    }
+}
+
+impl AsRef<str> for DnsName {
+    fn as_ref(&self) -> &str {
+        AsRef::<str>::as_ref(&self.0)
+    }
+}
+
+/// A type which encapsulates a borrowed string that is a syntactically valid DNS name.
+#[derive(Eq, Hash, PartialEq, Debug)]
+pub struct DnsNameRef<'a>(&'a str);
+
+impl<'a> DnsNameRef<'a> {
+    /// Copy this object to produce an owned `DnsName`.
+    pub fn to_owned(&'a self) -> DnsName {
+        DnsName(self.0.to_string())
+    }
+
+    /// Copy this object to produce an owned `DnsName`, smashing the case to lowercase
+    /// in one operation.
+    pub fn to_lowercase_owned(&'a self) -> DnsName {
+        DnsName(self.0.to_lowercase())
+    }
+}
+
+impl<'a> TryFrom<&'a str> for DnsNameRef<'a> {
+    type Error = InvalidDnsNameError;
+
+    fn try_from(value: &'a str) -> Result<DnsNameRef<'a>, Self::Error> {
+        validate(value.as_bytes())?;
+        Ok(DnsNameRef(value))
+    }
+}
+
+impl<'a> AsRef<str> for DnsNameRef<'a> {
+    fn as_ref(&self) -> &str {
+        self.0
+    }
+}
+
+/// "Labels must be 63 characters or less."
+const MAX_LABEL_LENGTH: usize = 63;
```

Comment:
So I would probably move `MAX_LABEL_LENGTH` below `State`, with `InvalidDnsNameError` between the two.

Reply:
thoughts on putting `MAX_LABEL_LENGTH` inside `fn validate`?

Reply:
Makes sense!